### PR TITLE
Add optional CPU profiler to Calypso server requests 

### PR DESCRIPTION
**Pull request by noahtallen**
#### Proposed Changes
With `USE_SERVER_PROFILER=true yarn start`, capture the CPU profile of individual routes in the Calypso server. This uses [`v8-profiler-next`](https://github.com/hyj1991/v8-profiler-next) to capture execution duration for individual JS functions.

Here's how it works:
- When the `USE_SERVER_PROFILER` environment variable exists, we add middleware to the express server which starts and stops the profiler.
- When starting the middleware, a new directory is created under `profiles` based on the current date/time. (So each execution of the server will have a different directory).
- When a request starts, the v8 profiler is also started. The profile name is set to the request URL. (with `/` replaced with `_`. So a request to `/themes` is named `_themes`)
- When the request is closed, the v8 profiler is stopped, and the profile is saved to a file with the `.cpuprofile` extension.
- Certain routes are ignored to prevent a lot of extra noise and to improve performance. (Like requests retrieving static files.)

The end result is that if you visit a single Calypso URL, a profile is saved which captures everything that happens as the server processes the request. **(Note that if you visit multiple URLs at the same time, several profiles will be saved, but each profile will contain the info from any other concurrent requests.)** So to get a one-time snapshot of a _single_ route's performance, you should _not_ have any concurrent requests.

#### Testing Instructions

1. Checkout the branch and run `yarn install`
2. Start the server with `USE_SERVER_PROFILER=true yarn start`
3. After the server boots and bundles everything, visit a calypso.localhost URL and wait for it to finish loading.
4. Inspect the `profiles` directory (in the top-level of Calypso). You should find a new directory for the date/time the server was started, and in it, you should find a file containing the profile. 
5. You should be able to open this file in VS Code to analyze the information. You can also install the extension VS Code will recommend to you to visualize the flamegraph:

image

image

### CODE REVIEW

File: client/server/pages/index.js [hunk 1]
```diff
@@ -52,20 +52,24 @@ const debug = debugFactory( 'calypso:pages' );
...
 
+let branchName;
 function getCurrentBranchName() {
-	try {
-		return execSync( 'git rev-parse --abbrev-ref HEAD' ).toString().replace( /\s/gm, '' );
-	} catch ( err ) {
-		return undefined;
+	if ( ! branchName ) {
+		try {
+			branchName = execSync( 'git rev-parse --abbrev-ref HEAD' ).toString().replace( /\s/gm, '' );
```
Comment (noahtallen) [inline client/server/pages/index.js]: These are drive-by local performance improvements. In dev mode, these will run on every request (and sometimes more frequently than that). `execSync` is relatively slow in this context, so caching the value while the server is online is a good way to speed things up.

File: client/server/boot/index.js [hunk 2]
```diff
@@ -25,6 +26,10 @@ export default function setup() {
 	app.use( userAgent.express() );
 	app.use( loggerMiddleware() );
 
+	if ( process.env.USE_PROFILER ) {
+		app.use( profilerMiddleware() );
```
Comment (tyxla) [inline client/server/boot/index.js]: Should we `require()` `profilerMiddleware` only if `USE_PROFILER` is enabled?

File: client/server/pages/index.js (hunk 1 above)
Comment (tyxla) [inline client/server/pages/index.js]: This is definitely a separate PR material. Nice optimization btw 👍

**Review by tyxla — COMMENTED** [review]
Thanks for this @noahtallen. 

I've left a few comments, LMK what you think.

What's with the failing e2e tests?

File: client/server/boot/index.js (hunk 2 above)
Comment (noahtallen) [reply client/server/boot/index.js]: Yeah, that's a great idea!

File: client/server/pages/index.js (hunk 1 above)
Comment (noahtallen) [reply client/server/pages/index.js]: Sure thing 👍

**Comment by noahtallen** [comment]
Unsure! I'll be investigating soon, as they shouldn't all be failing like that.

File: client/server/boot/index.js (hunk 2 above)
Comment (noahtallen) [reply client/server/boot/index.js]: Done!

File: client/server/pages/index.js (hunk 1 above)
Comment (noahtallen) [reply client/server/pages/index.js]: https://github.com/Automattic/wp-calypso/pull/66517

**Comment by noahtallen** [comment]
> What's with the failing e2e tests?

I guess importing the profiler at all in production must have caused those issues? Now that we only import dynamically, it looks like all is well.

File: client/server/middleware/profiler.js [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+ * This should be dynamically imported:
+ *
+ * if ( shouldProfile ) {
+ *   app.use ( require( 'calypso/server/middleware/profiler' )() )
+ * }
+ *
+ * to avoid importing v8-profiler-next in production environments.
+ */
+module.exports = () => {
+	let IS_PROFILING = false;
```
Comment (tyxla) [inline client/server/middleware/profiler.js]: Should we log something in the console that indicates that we're running Calypso with a profiler enabled?

File: client/server/middleware/profiler.js [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+		res.on( 'close', () => {
+			IS_PROFILING = false;
+			const profile = v8Profiler.stopProfiling( profileName );
+
+			profile.export( ( error, result ) => {
+				if ( error ) {
+					console.error( error );
+					return;
+				}
+				fs.writeFile( `${ profilesRoot }/${ profileName }.cpuprofile`, result, ( err ) => {
```
Comment (tyxla) [inline client/server/middleware/profiler.js]: Should we log something in the console on a successful profile write?

File: client/server/boot/index.js [hunk 5]
```diff
@@ -25,6 +25,10 @@ export default function setup() {
 	app.use( userAgent.express() );
 	app.use( loggerMiddleware() );
 
+	if ( process.env.USE_PROFILER === 'true' ) {
```
Comment (tyxla) [inline client/server/boot/index.js]: Could we document that somewhere? Perhaps either in the `development-workflow.md` document or in the `performance.md` one?

Comment (tyxla) [inline client/server/boot/index.js]: Should we make it clear in the feature flag name that this profiler is just for the server side?

File: client/server/middleware/profiler.js [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+ * to avoid importing v8-profiler-next in production environments.
+ */
+module.exports = () => {
+	let IS_PROFILING = false;
+
+	// Generates a CPU profile compatible with VS Code's viewer.
+	v8Profiler.setGenerateType( 1 );
+
+	const profilesRoot = path.resolve(
+		path.join( __dirname, '../../../profiles', new Date().toISOString() )
```
Comment (tyxla) [inline client/server/middleware/profiler.js]: I wonder if we should group the profiles by profile name (so the folder would be named with the profile name) and then by date - so we could easily compare 2 profiles of the same request without having to lookup through the profiles thoroughly.

This will also help against overfilling the `profiles` directory with directories if we often work with profiling enabled. 

Both ways have trade-offs, so I just wanted to hear what you think about this.

File: client/server/middleware/profiler.js [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+	return ( req, res, next ) => {
+		// Avoid profiling certain requests (like for static files) and don't
+		// start profiling if we already are.
+		if (
+			IS_PROFILING ||
+			! req.originalUrl.startsWith( '/' ) ||
+			req.originalUrl.startsWith( '/calypso/' ) ||
+			req.originalUrl.startsWith( '/service-worker' ) ||
+			req.originalUrl.startsWith( '/nostats.js' ) ||
+			req.originalUrl.startsWith( '/version' )
```
Comment (tyxla) [inline client/server/middleware/profiler.js]: I think we should also disable `__webpack_hmr` here. You'll get such a profile if you try to access Calypso while it's building the client.

File: client/server/middleware/profiler.js (hunk 4 above)
Comment (tyxla) [inline client/server/middleware/profiler.js]: For some reason, this always ends up being named `_.cpuprofile` for me, and sometimes `___webpack_hmr.cpuprofile`. Perhaps `req.originalUrl` isn't always populated as we expect?

File: client/server/middleware/profiler.js (hunk 3 above)
Comment (tyxla) [inline client/server/middleware/profiler.js]: Naming nit: we usually use all caps and underscore case for constants, but this is not the case here. Should we just conform to our regular variable name conventions?

File: client/server/middleware/profiler.js [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+		fs.accessSync( profilesRoot );
+	} catch {
+		fs.mkdirSync( profilesRoot, { recursive: true } );
+	}
+
+	return ( req, res, next ) => {
+		// Avoid profiling certain requests (like for static files) and don't
+		// start profiling if we already are.
+		if (
+			IS_PROFILING ||
```
Comment (tyxla) [inline client/server/middleware/profiler.js]: What's the purpose behind `IS_PROFILING` in the first place? To prevent profiling multiple concurrent requests?

File: client/server/middleware/profiler.js (hunk 4 above)
Comment (noahtallen) [reply client/server/middleware/profiler.js]: This actually makes sense -- I just chose `_` to replace `/` since `/` is also used by the filesystem. So visiting `/` outputs `_`, visiting `/themes/foo` outputs `_themes_foo`. So in your case, that's just the base URL of Calypso.

Let me know if you have any better ideas!

File: client/server/middleware/profiler.js (hunk 3 above)
Comment (noahtallen) [reply client/server/middleware/profiler.js]: Hm, I understand what you mean. I always thought all caps was useful for "global" variables, which is kind of how I was treating this... just scoped more within the function. (In a different approach, I had it in the top-level of the file.)

I can certainly change it, though.

File: client/server/middleware/profiler.js (hunk 8 above)
Comment (noahtallen) [reply client/server/middleware/profiler.js]: Right, so if you visit "/themes", it will start profiling. If you then also visit "/home/foo" before the first request finishes, it would start profiling again with a different ID. At the end, you would get two different profiles with overlapping data, since obviously the CPU is shared between the two requests.

Since my main goal was to profile a single request, I figured it would be better to just start the profiler once. (And if you were to visit a second URL concurrently, that data is still included in that first profile, up to the point the first request finishes.)

File: client/server/middleware/profiler.js (hunk 6 above)
Comment (noahtallen) [reply client/server/middleware/profiler.js]: I like this idea 👍

File: client/server/boot/index.js (hunk 5 above)
Comment (noahtallen) [reply client/server/boot/index.js]: Sure -- here are some ideas:

- `USE_REQUEST_PROFILER`
- `USE_EXPRESS_PROFILER`
- `USE_ROUTE_PROFILER`
- `USE_SERVER_REQUEST_PROFILER`

File: client/server/middleware/profiler.js (hunk 3 above)
Comment (tyxla) [reply client/server/middleware/profiler.js]: Yeah, it's minor but I'd suggest changing that if you don't mind 😉

File: client/server/middleware/profiler.js (hunk 4 above)
Comment (tyxla) [reply client/server/middleware/profiler.js]: Thanks for elaborating, I think it makes sense indeed 👍

File: client/server/boot/index.js (hunk 5 above)
Comment (tyxla) [reply client/server/boot/index.js]: `USE_SERVER_REQUEST_PROFILER` or even `USE_SERVER_PROFILER` work well for me 👍

File: client/server/middleware/profiler.js (hunk 8 above)
Comment (tyxla) [reply client/server/middleware/profiler.js]: Makes sense. I think it's worth documenting this trait as well to avoid confusion 😉

**Comment by noahtallen** [comment]
Alright, all feedback should be addressed!

File: client/server/pages/index.js [hunk 9]
```diff
@@ -52,24 +52,20 @@ const debug = debugFactory( 'calypso:pages' );
...
-	return branchName;
 }
 
-let commitChecksum;
 function getCurrentCommitShortChecksum() {
-	if ( ! commitChecksum ) {
-		try {
-			commitChecksum = execSync( 'git rev-parse --short HEAD' ).toString().replace( /\s/gm, '' );
-		} catch ( err ) {}
+	try {
```
Comment (tyxla) [inline client/server/pages/index.js]: Hm, why did we revert these?

**Review by tyxla — APPROVED** [review]
This is looking great. Let's ship it! 🚀  Thanks!

File: client/server/pages/index.js (hunk 9 above)
Comment (noahtallen) [reply client/server/pages/index.js]: Great question -- I guess a rebase issue? (I did have a commit in this PR reverting the change so we could move it somewhere else)
